org/mozilla/jss/ssl/SSLSocket.java

```diff
@@ -547,16 +548,76 @@ public SSLSocket(InetAddress address, int port,
         SSLClientCertificateSelectionCallback clientCertSelectionCallback)
             throws IOException
     {
-        this(address, address.getHostName(), port, localAddr, localPort,
+        this(address, address.getHostName(), port, localAddr, localPort, null /* alpn */,
             certApprovalCallback, clientCertSelectionCallback);
     }
 
+    /**
+     * Create an SSL client socket and connects to the specified
+     * address and port. Installs the given callbacks for
+     * certificate approval and client certificate selection.
+     * Supports Application Layer Protocol Negotiation (ALPN)
+     * extension.
+     *
+     * @param address The IP address to connect to.
+     * @param port The port to connect to.
+     * @param hostname Hostname to use for SNI
+     * @param alpn Array of ALPN protocol IDS.  If empty, extension is not
+     *      added.  Each protocol ID must have a length between 1 and 255
+     *      bytes.
+     * @param certApprovalCallback A callback that can be used to override
+     *      approval of the peer's certificate.
+     * @param clientCertSelectionCallback A callback to select the client
+     *      certificate to present to the peer.
+     */
+    public SSLSocket(
+        InetAddress address, int port,
+        String hostname, byte[][]alpn,
```

Comment:
In your [commit message](https://github.com/frasertweedale/jss/commit/465ac1d9b679334d9531c343988e6609e12586e9) you wrote:

`   Update the private SSLSocket constructor to accept a String[] of ALPN`

But I only see `byte[][]` of ALPNs. I'd prefer we stick with JCA-standard types here rather than introducing our own in general... But at minimum, the commit message should match the code. :-)

Reply:
No, it should be `byte[]` - this is how ALPN protocol IDs are defined.  See https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids and https://tools.ietf.org/html/rfc7301#section-6 .

Reply:
Just saw your comment on the other PR.  Bravo Java, bravo.  What if the sequence of bytes is not valid UTF-8?  What do you know, there are ALPN GREASE values that are not valid UTF-8 (see https://tools.ietf.org/html/rfc8701, e.g. `{0xFA, 0xFA}`.  If you want to set them, you cannot.  If you want to read them, you cannot (the UTF-8 reader converts them to `��`.

Reply:
:-)